api/src/main/java/marquez/db/OpenLineageDao.java

```diff
@@ -745,16 +750,58 @@ default Set<DatasetId> toDatasetId(List<Dataset> datasets) {
 
   default void updateMarquezOnComplete(
       LineageEvent event, UpdateLineageRow updateLineageRow, RunState runState) {
+    final JobVersionDao jobVersionDao = createJobVersionDao();
+    // Link the job version to the job only if the run is marked done and has transitioned into one
+    // of the following states: COMPLETED, ABORTED, or FAILED.
+    final boolean linkJobToJobVersion = runState.isDone();
+
     BagOfJobVersionInfo bagOfJobVersionInfo =
-        createJobVersionDao()
-            .upsertJobVersionOnRunTransition(
-                updateLineageRow.getJob(),
-                updateLineageRow.getRun().getUuid(),
-                runState,
-                event.getEventTime().toInstant());
+        jobVersionDao.upsertJobVersionOnRunTransition(
+            jobVersionDao.loadJobRowRunDetails(
+                updateLineageRow.getJob(), updateLineageRow.getRun().getUuid()),
+            runState,
+            event.getEventTime().toInstant(),
+            linkJobToJobVersion);
     updateLineageRow.setJobVersionBag(bagOfJobVersionInfo);
   }
 
+  /**
+   * A separate method is used as the logic to update Marquez model differs for streaming and batch.
+   * The assumption for batch is that the job version is created when task is done and cumulative
+   * list of input and output datasets from all the events is used to compute the job version UUID.
+   * However, this wouldn't make sense for streaming jobs, which are mostly long living and produce
+   * output before completing.
+   *
+   * <p>In this case, a job version is created based on the list of input and output datasets
+   * referenced by this job. If a job starts with inputs:{A,B} and outputs:{C}, new job version is
+   * created immediately at job start. If a following event produces inputs:{A}, outputs:{C}, then
```

Comment:
If a job has `{A,B}` as initial inputs, the **only** `{A}` shouldn't that be a new job version?

Reply:
Existing algorithm to compute version of the job relies on all the inputs & datasets for the particular run. We should not modify it, as this would cause new job version for all the existing jobs, but we could create a separate version to evaluate version of the streaming job if we wanted. 

However, when looking into the approach, I found it useful. It's a cumulative approach, where new job version is created if a new input/output dataset is involved in processing. If some dataset was included in the past events, but is no longer present, the version does not get change. 

The benefit of this assumption is that we don't require producer to emit all the datasets all the time. If you emit amount of bytes written into output dataset, without containing input dataset in the event, it doesn't mean there is new job version without the inputs. 

Is it OK?

Reply:
I do feel if a stream is removed, you'll want to remove that edge from the _current_ lineage graph. But, I also understand the limitations here, as you mentioned, with bytes written to an _output_ dataset present with no _input_ datasets specified; that said, and after thinking it through, this may be a noop and the logic you have is reasonable.

For example, let's say we have a streaming job `X` with inputs `{A,B}` and outputs `C`. The job runs and is marked as `RUNNING` with the run ID `74f117af-eb90-4ffd-98e1-c1bc215934df`. To change the inputs from `{A,B}` to `{B}` (or similarly for the outputs), the user will have to redeploy the job (with new code possibly) and therefore be associated with a new run ID. So, what I think you have is logically correct given how streaming jobs are deployed. For batch jobs, versioning is more straight forward as we version from run-to-run.